### PR TITLE
Progress TypeScript migration

### PR DESCRIPTION
**Pull request by jwhitham-ds**
I've picked up from the starts that @Scalahansolo and @zhouzi have made. I was progressing through writing types out on another branch, but I thought that it would be good make sure that this is something that is still supported and to raise an early PR.

This PR just makes filename changes, moves devDependencies up to the root of the repo and modifies some config. Once it is approved I'll continue working on getting everything typed properly.

This branch is rebased on the latest upstream main and `yarn && yarn build` runs fine without errors or warnings.

image

### CODE REVIEW

**Comment by jwhitham-ds** [comment]
Hi guys, I ran into issues trying to get Jest set up for TypeScript and then circumstances changed at work and I couldn't spend any more time on this. I hope to spend some of my own time on it this weekend, but if anyone feels like jumping in and helping out that I would welcome that.

**Comment by jwhitham-ds** [comment]
Turns out I was most of the way there with my WIP and Jest and I just had to remove the `--experimental-vm-modules` option to get the tests working =D.

With that option enabled, I was getting the following error:
image

**Comment by TranquilMarmot** [comment]
What's the status of this? Need help on any parts of it?

**Comment by jwhitham-ds** [comment]
This is just waiting for review

File: packages/emoji-mart/tsconfig.json
Comment (JoshuaKGoldberg) [inline packages/emoji-mart/tsconfig.json:16]: TypeScript doesn't perform type checking on `.scss` files. There's no real need to include them explicitly.

```suggestion
  "include": ["src"],
```

File: packages/emoji-mart/tsconfig.json
Comment (JoshuaKGoldberg) [inline packages/emoji-mart/tsconfig.json:8]: This repo doesn't use them and I wouldn't recommend enabling them without a need. TC39 is going to make a different version any .. year now.

```suggestion
```

File: packages/emoji-mart/tsconfig.json
Comment (JoshuaKGoldberg) [inline packages/emoji-mart/tsconfig.json:3]: All modern evergreen browsers & runtimes support at least ES2020 now. No need to transpile down beyond it.

```suggestion
    "target": "es2020",
```

File: jest.config.js
Comment (JoshuaKGoldberg) [inline jest.config.js:179]: `tsx?` means `"ts"`, and optionally `"x"`. `ts?` means `"t"`, and optionally `"s"`. There's no need for the latter.

```suggestion
    '\\.tsx?$': 'ts-jest',
```

File: packages/emoji-mart/src/browser.js
Comment (JoshuaKGoldberg) [inline packages/emoji-mart/src/browser.js:1]: Fun fact: the newest versions of TypeScript do now support `.js` imports! https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#ecmascript-module-support-in-nodejs

Including `.js` in the file imports is fine now. So all these `.js` removals can safely be reverted.

```suggestion
import * as EmojiMart from './index.js'
```

**Review by JoshuaKGoldberg — COMMENTED** [review]
Hi hi! I'm not a maintainer on this repo, so there's no need to listen to my review. Just popping in out of enthusiasm for this PR landing, and to try to be helpful. 🙂

File: package.json
Comment (JoshuaKGoldberg) [inline package.json:16]: Running `yarn test` from the root gives a bunch of:

```
ts-jest[config] (WARN) message TS151001: If you have issues related to imports, you should consider setting `esModuleInterop` to `true` in your TypeScript configuration file (usually `tsconfig.json`). See https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/#easier-ecmascript-module-interoperability for more information.
```

Perhaps adding a `tsconfig.json` similar to `./packages/emoji-mart/tsconfig.json` to the root of the project is a good idea? I'd also suggest adding `"noEmit": true` to both, since they're just used for IDE checking & in Jest.

**Review by EtienneLem — APPROVED** [review]

**Pull request merged**
